### PR TITLE
✨ add a fancy GitHub Action for 'shellcheck'

### PR DESCRIPTION
**Pull request by thomasmerz**
I have promised in https://github.com/yubiuser/pihole_adlist_tool/pull/37#issuecomment-1025232709  that I will make a PR for `shellcheck` - here we go 😁

### CODE REVIEW

**Comment by thomasmerz** [comment]
Here are my tests, so you can see what this action is doing and under which condition it is doing what: https://github.com/thomasmerz/pihole_adlist_tool/actions

image

**Comment by thomasmerz** [comment]
An E-Mail will be sent in case of an error with link to the action that failed:
image

File: shellcheck.yml [hunk 1]
```diff
@@ -0,0 +1,43 @@
+name: Shellcheck Lint
+
+on:
+  push:
+    paths:
```
Comment (yubiuser) [inline shellcheck.yml:5]:
```suggestion

```
I would remove that. I'm happy if it would run on all paths and all branches.

File: shellcheck.yml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      # Run workflow on push to file pihole_adlist_tool
+      - pihole_adlist_tool
+
+  pull_request:
+    branches:
+      # Run workflow on pull request to master
+      - master
```
Comment (yubiuser) [inline shellcheck.yml:12]: I would remove that. I'm happy if it would run on all paths and all branches.
```suggestion

```

File: shellcheck.yml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+      # Download Shellcheck and add it to the workflow path
+      - name: Download Shellcheck
+        run: |
+          wget -qO- "https://github.com/koalaman/shellcheck/releases/download/stable/shellcheck-stable.linux.x86_64.tar.xz" | tar -xJv
+          chmod +x shellcheck-stable/shellcheck
+      # Verify that Shellcheck can be executed
+      - name: Check Shellcheck Version
+        run: |
+          shellcheck-stable/shellcheck --version
```
Comment (yubiuser) [inline shellcheck.yml:36]: Don't we need a if clause here to test if there is any output?

**Review by yubiuser — COMMENTED** [review]
Thank you very much for this PR. Only downside for me - I didn't learn how to do it myself ;-)

As written in the comments, I'd prefer to have it running in all pushes and branches.

**Comment by yubiuser** [comment]
From the documentation

> You must store workflow files in the `.github/workflows` directory of your repository.

File: shellcheck.yml (hunk 3 above)
Comment (thomasmerz) [reply shellcheck.yml:36]: No, we don't need an `if` clause or something like that. Have a look at the link to my fork, where you can see the output from one failing example (I didn't ignore SC2166 for this run):
https://github.com/thomasmerz/pihole_adlist_tool/runs/4999213693?check_suite_focus=true

File: shellcheck.yml (hunk 2 above)
Comment (thomasmerz) [reply shellcheck.yml]: This won't work. We need this instead:
```
on:
  push:
    branches:
      # Run workflow on every push
      - '**'        # matches every branch

  pull_request:
    branches:
      # Run workflow on every push
      - '**'        # matches every branch
```

File: shellcheck.yml (hunk 1 above)
Comment (thomasmerz) [reply shellcheck.yml]: This won't work. We need this instead:
```
on:
  push:
    branches:
      # Run workflow on every push
      - '**'        # matches every branch

  pull_request:
    branches:
      # Run workflow on every push
      - '**'        # matches every branch
```

**Comment by thomasmerz** [comment]
@yubiuser , I adapted your recommendations in my commit/PR :)

**Comment by thomasmerz** [comment]
Sorry, for "stealing" the fun 😉 But maybe you might have some fun by learning and showing me, how this should look like using the official docker-container from koalaman `koalaman/shellcheck:stable` as I showed you how this works on GitLab?!

**Comment by thomasmerz** [comment]
> You must store workflow files in the `.github/workflows` directory of your repository.

Funny that it even works/worked if workflow files are stored outside the `.github/workflows` directory 😜

**Comment by thomasmerz** [comment]
You could/should also **fix some repo-settings** to benefit of the shellcheck action:

image

**Pull request merged**